gateway/gateway.go

```diff
@@ -131,7 +159,7 @@ type IPFSBackend interface {
 	// that may contain a single error for if any errors occur during the streaming. If there was an initial error the
 	// error channel is nil
 	// TODO: Make this function signature better
-	GetCAR(context.Context, ImmutablePath) (ContentPathMetadata, io.ReadCloser, <-chan error, error)
+	GetCAR(context.Context, ImmutablePath, CarParams) (io.ReadCloser, <-chan error, error)
```

Comment:
This is an interface breaking change which was needed to accept the CAR params anyway given that we likely don't want to add any more confusion to the different "path" types until we can unify them better.

I also dropped the return of `ContentMetadataPath` since we now return all the blocks in the path instead of skipping the path blocks.

TBD: While we're breaking this anyway we should probably better handle the error signatures here. Suggestions welcome.

Reply:
@aschmahmann can we remove `<-chan error`, wrap `io.ReadCloser` and add the error there (on `.Close` perhaps, if multiple errors, send a multierror)?

Reply:
Ok, I checked the code more carefully. This should be doable and it makes sense, since the channel is for streaming errors. What's better for that the the Reader?

Reply:
Maybe we can somehow use pipe's `CloseWithError` for this?

Reply:
Yes, we can. Pipe's writer.Close never returns an error if we look at the source code. So using that is useless.

Reply:
So, I've done some digging. Using `io.PipeWriter` is very nice for error handling! `io.PipeWriter.Close` calls `io.PipeWriter.CloseWithError(nil)` which always... returns `nil`: https://pkg.go.dev/io#PipeWriter.CloseWithError

We're checking for the error from `w.Close()` which is guaranteed to always be non-existing. We can do is call `CloseWithError` directly with the error we want, which will make subsequent reads from the pipe return the error we specified. Therefore removing the need for our wonky channel.

Reply:
@aschmahmann see my commit ([6520009](https://github.com/ipfs/boxo/pull/303/commits/652000913d9bcfa66156bfae4553a51a4968655a)). Feel free to discard it if it has problems.